addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FullPathDisclosureAlertPassiveScanRule.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.pscanrulesAlpha;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+public class FullPathDisclosureAlertPassiveScanRule extends PluginPassiveScanner {
+    private final int pluginID = 110009;
+    private static final String MESSAGE_PREFIX = "pscanalpha.fullpathdisclosurealert.";
+
+    @Override
+    public int getPluginId() {
+        return this.pluginID;
+    }
+
+    @Override
+    public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+        final int statusCode = msg.getResponseHeader().getStatusCode();
+        final String responseBody = msg.getResponseBody().toString();
+        String evidence;
+        // Matches the presence of one of the words in the response body
+        Pattern errorPattern = Pattern.compile("(?i)error|parse|fatal|warning|notice");
+        Matcher errorMatcher = errorPattern.matcher(responseBody);
+        // matches the presence of either windows style or unix style full paths
+        // example for windows : C:\folder\folder
+        // example for unix : /dir/dir
+        Pattern pathPattern =
+                Pattern.compile("(?i)(/([a-z0-9]+/)+)|(([a-z0-9]:\\\\)+([a-z0-9]+\\\\)+)");
+        Matcher pathMatcher = pathPattern.matcher(responseBody);
+        if (!(statusCode < 200 || statusCode > 299)) return;
+        if (!(errorMatcher.find())) return;
```

Comment:
These should use curly braces for consistency.
Instead of checking status code you should use isPageSuccess. Though I'm not sure I agree with skipping responses for either of these reasons.

Reply:
consider this HATEOAS response
`HTTP/1.1 200 OK
Content-Type: application/+json
Content-Length: ...
{
   "payroll": {
       "employee_number": "employee_123",
       "salary" : 1000,
       "links": {
           "increment": "/payroll/employee_123/increment",
           "decrement": "/payroll/employee_123/decrement",
           "close": "/payroll/employee_123/close"
       }
   }
}`
Here `"/payroll/employee_123/increment"` is matched to a unix based file path by regex.
Here is another example that leads to a false positive `<a href=https://example.com/somelink/path >click me</a>`
Thus to avoid false positives, I have introduced these checks.

Reply:
Neither of these checks would address those examples.

It's a 200 so it passes, it doesn't have one of the error words.

Further a redirect 3xx, client error 4xx, or server error 5xx could definitely disclose a full path. In fact IIS is almost guaranteed to on 5xx when custom error pages aren't enabled.

Maybe instead of trying to exclude error words include some default linux and macOS paths?
/bin
/usr
/mnt
/proc
/sbin
/dev
/lib
/tmp
/opt
/home
/var
/root
/etc
(☝ not necessarily ordered in any particularly useful way)

/Applications
/Volumes
/System
/Users
/Developer
/Library
(☝ not necessarily ordered in any particularly useful way)

You can easily add a UnitTest to try :wink:

Reply:
The inclusion of default linux and windows paths in a vulnerable request is a good idea and I will implement it replacing the errorMatcher. However, I am a little confused with the statusCode.

So this response, being a success and not containing any FPD vulnerabilities should not raise any alerts.

```
HTTP/1.1 200 OK 
Content-Type: application/+json 
 Content-Length: ... 
{ "payroll": { "employee_number": "employee_123", "salary" : 1000, "links": { "increment": "/payroll/employee_123/increment", "decrement": "/payroll/employee_123/decrement", "close": "/payroll/employee_123/close" } } }
```

So the line, (taking `statusCode` as `200`)
`if (!(statusCode < 200 || statusCode > 299)) return;` evaluates to `if(!(false || false))` => `if( !(false) )` => `if(true)` will cause `return` to be hit. Thus, no alerts raised.
Only if the response did not succeed (that is 1xx or 3xx or 4xx or 5xx) will it continue to the lines after the guard statement where we check for default vulnerable paths and perform regex pattern matching to extract the found path.

> Neither of these checks would address those examples.

> It's a 200 so it passes, it doesn't have one of the error words.

My understanding contradicts this. Am I missing something here?

I wrote these two unit tests and both of them passed.

```
void sampleTest1() throws URIException{
        //Given
        HttpMessage msg = createMessage("{ \"payroll\": { \"employee_number\": \"employee_123\", \"salary\" : 1000, \"links\": { \"increment\": \"/payroll/employee_123/increment\", \"decrement\": \"/payroll/employee_123/decrement\", \"close\": \"/payroll/employee_123/close\" } } }\n" +
                "Here \"/payroll/employee_123/increment\"" , HttpStatusCode.OK);
        //When
        scanHttpResponseReceive(msg);
        //Then
        assertThat(alertsRaised , hasSize(0));
    }
```

```
void sampleTest2() throws URIException{
        //Given
        HttpMessage msg = createMessage("<a href=\"https://example.com/foo/bar/\">" , HttpStatusCode.OK);
        //When
        scanHttpResponseReceive(msg);
        //Then
        assertThat(alertsRaised , hasSize(0));
    }
```

Reply:
That’s my fault, I was ignoring the leading bang/not (!).

Reply:
So I will preserve the statusCheck but make it cleaner by moving the it to a separate private function.

Reply:
No you should use `!isSuccess`

https://github.com/zaproxy/zaproxy/blob/8e78a891d88aeadb6c6007cb2437982eb73536ef/zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractPlugin.java#L747